File: Tween/ListElement.vb
' Tween - Client of Twitter
' Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
'           (c) 2008-2011 Moz (@syo68k)
'           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
'           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
'           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
' All rights reserved.
' 
' This file is part of Tween.
' 
' This program is free software; you can redistribute it and/or modify it
' under the terms of the GNU General Public License as published by the Free
' Software Foundation; either version 3 of the License, or (at your option)
' any later version.
' 
' This program is distributed in the hope that it will be useful, but
' WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
' or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
' for more details. 
' 
' You should have received a copy of the GNU General Public License along
' with this program. If not, see <http://www.gnu.org/licenses/>, or write to
' the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
' Boston, MA 02110-1301, USA.

Public Class ListElement
    Public Id As Long = 0
    Public Name As String = ""
    Public Description As String = ""
    Public Slug As String = ""
    Public IsPublic As Boolean = True
    Public SubscriberCount As Integer = 0   '購読者数
    Public MemberCount As Integer = 0   'リストメンバ数
    Public UserId As Long = 0
    Public Username As String = ""
    Public Nickname As String = ""

    Protected _tw As Twitter

    Private _members As List(Of UserInfo) = Nothing
    Private _cursor As Long = -1

    Public Sub New()

    End Sub

    Public Sub New(ByVal listElementData As TwitterDataModel.ListElementData, ByVal tw As Twitter)
        Me.Description = listElementData.Description
        Me.Id = listElementData.Id
        Me.IsPublic = (listElementData.Mode = "public")
        Me.MemberCount = listElementData.MemberCount
        Me.Name = listElementData.Name
        Me.SubscriberCount = listElementData.SubscriberCount
        Me.Slug = listElementData.Slug
        Me.Nickname = listElementData.User.Name.Trim()
        Me.Username = listElementData.User.ScreenName
        Me.UserId = listElementData.User.Id

        Me._tw = tw
    End Sub

    Public Overridable Function Refresh() As String
        Return _tw.EditList(Me.Id.ToString(), Name, Not Me.IsPublic, Me.Description, Me)
    End Function

    <Xml.Serialization.XmlIgnore()>
    Public ReadOnly Property Members As List(Of UserInfo)
        Get
            If Me._members Is Nothing Then Me._members = New List(Of UserInfo)
            Return Me._members
        End Get
    End Property

    <Xml.Serialization.XmlIgnore()>
    Public ReadOnly Property Cursor As Long
        Get
            Return _cursor
        End Get
    End Property

    Public Function RefreshMembers() As String
        Dim users As New List(Of UserInfo)()
        _cursor = -1
        Dim result As String = Me._tw.GetListMembers(Me.Id.ToString(), users, _cursor)
        Me._members = users
        Return If(String.IsNullOrEmpty(result), Me.ToString, result)
    End Function

    Public Function GetMoreMembers() As String
        Dim result As String = Me._tw.GetListMembers(Me.Id.ToString(), Me._members, _cursor)
        Return If(String.IsNullOrEmpty(result), Me.ToString, result)
    End Function

    Public Overrides Function ToString() As String
        Return "@" + Username + "/" + Name + " [" + If(Me.IsPublic, "Public", "Protected") + "]"
    End Function
End Class

File: OpenTween/ListElement.cs
// OpenTween - Client of Twitter
// Copyright (c) 2007-2011 kiri_feather (@kiri_feather) <kiri.feather@gmail.com>
//           (c) 2008-2011 Moz (@syo68k)
//           (c) 2008-2011 takeshik (@takeshik) <http://www.takeshik.org/>
//           (c) 2010-2011 anis774 (@anis774) <http://d.hatena.ne.jp/anis774/>
//           (c) 2010-2011 fantasticswallow (@f_swallow) <http://twitter.com/f_swallow>
//           (c) 2011      Egtra (@egtra) <http://dev.activebasic.com/egtra/>
// All rights reserved.
//
// This file is part of OpenTween.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>, or write to
// the Free Software Foundation, Inc., 51 Franklin Street - Fifth Floor,
// Boston, MA 02110-1301, USA.

#nullable enable

using System.Collections.Generic;
using System.Threading.Tasks;
using System.Xml.Serialization;
using OpenTween.Api.DataModel;

namespace OpenTween
{
    public class ListElement
    {
        public long Id = 0;
        public string Name = "";
        public string Description = "";
        public string Slug = "";
        public bool IsPublic = true;
        public int SubscriberCount = 0;   // 購読者数
        public int MemberCount = 0;   // リストメンバ数
        public long UserId = 0;
        public string Username = "";
        public string Nickname = "";

        protected Twitter tw = null!;

        private List<UserInfo> members = new();

        [XmlIgnore]
        public long Cursor { get; private set; } = -1;

        public ListElement()
        {
        }

        public ListElement(TwitterList listElementData, Twitter tw)
        {
            this.Description = listElementData.Description;
            this.Id = listElementData.Id;
            this.IsPublic = listElementData.Mode == "public";
            this.MemberCount = listElementData.MemberCount;
            this.Name = listElementData.Name;
            this.SubscriberCount = listElementData.SubscriberCount;
            this.Slug = listElementData.Slug;
            this.Nickname = listElementData.User.Name.Trim();
            this.Username = listElementData.User.ScreenName;
            this.UserId = listElementData.User.Id;

            this.tw = tw;
        }

        public virtual async Task Refresh()
        {
            var newList = await this.tw.EditList(this.Id, this.Name, !this.IsPublic, this.Description)
                .ConfigureAwait(false);

            this.Description = newList.Description;
            this.Id = newList.Id;
            this.IsPublic = newList.IsPublic;
            this.MemberCount = newList.MemberCount;
            this.Name = newList.Name;
            this.SubscriberCount = newList.SubscriberCount;
            this.Slug = newList.Slug;
            this.Nickname = newList.Nickname;
            this.Username = newList.Username;
            this.UserId = newList.UserId;
        }

        [XmlIgnore]
        public List<UserInfo> Members
            => this.members;

        public async Task RefreshMembers()
        {
            var users = new List<UserInfo>();
            this.Cursor = await this.tw.GetListMembers(this.Id, users, cursor: -1)
                .ConfigureAwait(false);
            this.members = users;
        }

        public async Task GetMoreMembers()
            => this.Cursor = await this.tw.GetListMembers(this.Id, this.members, this.Cursor)
                .ConfigureAwait(false);

        public override string ToString()
            => $"@{this.Username}/{this.Name} [{(this.IsPublic ? "public" : "Protected")}]";
    }
}
